tag/template_service.go

```diff
@@ -0,0 +1,158 @@
+package tag
+
+import (
+	"fmt"
+
+	"github.com/odpf/columbus/tag/validator"
+	"github.com/pkg/errors"
+)
+
+var validDomainFieldDataType = []string{
+	"string",
+	"double",
+	"boolean",
+	"enumerated",
+	"datetime",
+}
+
+// Service is a type of service that manages business process
+type TemplateService struct {
+	validator  validator.Validator
+	repository TemplateRepository
+}
+
+// Validate validates domain template based on the business rule
+func (s *TemplateService) Validate(template Template) error {
+	err := s.validator.Validate(template)
+	if err != nil {
+		err = ValidationError{err}
+	}
+
+	return err
+}
+
+// Create handles create business operation for template
+func (s *TemplateService) Create(template *Template) error {
+	if template == nil {
+		return errors.New("template is nil")
+	}
+	err := s.Validate(*template)
+	if err != nil {
+		return err
+	}
+
+	filterForExistence := Template{
+		URN: template.URN,
+	}
+	templateRecords, err := s.repository.Read(filterForExistence)
+	if err != nil {
+		return errors.Wrap(err, "error checking template existence")
+	}
+	if len(templateRecords) > 0 {
```

Comment:
If Read always expect 1 result, then why not doing this validation inside the read itself?

Reply:
checking from Dexter code, it seems like they design `Read` to return list.
It is also being used to fetch list of templates in this file on `Index` method.